### DISCUSSION TITLE
Hashtable with key and value as strings

### DISCUSSION THREAD

**Discussion by shwetaiisc** [NVIDIA/cuCollections#303] [Q&A] [answered]
Hi, 

I have been trying to implement a cuco hashtable with the following requirements --
1. The key is a string (char*), the value is also a string (char*) and an access counter (HT<char*, {char*, int}>). 
2. The CPU inserts all key, value pairs, and sets the access counter 0. 
3. The GPU later reads these keys, returns the corresponding string values, and increments the access counter. 

I am facing a lot of challenges in implementing this. I wanted to know if it is possible to achieve the above requirements using cuco.   

Thanks!

**Comment by PointKernel** [comment, marked as answer]
> The key is a string (char*), the value is also a string (char*) and an access counter (HT<char*, {char*, int}>).

String is not a fixed-size data type thus would be really inefficient to perform hash table operations like insertion or query. Just to share some ideas with you, for an array of string data, we often store array indices as hash map keys/values thus the map is more of a mapping table. This also requires some extra work, e.g., you need to pass a custom hasher and a custom key equal functor to the map to make it work.

> The CPU inserts all key, value pairs, and sets the access counter 0.

It's possible, but I highly recommend not to do that. cuco hash table is GPU-only

**Reply by shwetaiisc** [reply]
Thanks for the reply. I ended up doing the same. 
I took the string and hashed it to an 8-byte integer, and the value helped me index into a separate buffer that stores the actual key-value pairs.
I want to implement a least frequently used replacement policy in the hashmap when the hashmap reaches a certain threshold. 
Any recommendations for how to achieve it?

**Reply by PointKernel** [reply]
The LFU cache requires "orderness" while a hash map is an unordered data structure. Not sure how to do it efficiently.

Yours to discover. :)
